### ISSUE TITLE
How should I optimize children aggregation usage?

### ISSUE THREAD

**Issue by iwhitfield** [elastic/elasticsearch#13113]
It appears that the children aggregation is not cached (correct term?) using v1.6, as it is consistently taking the same amount of time for my query.

Here's an example of my aggs:

```
{
    "aggs": {
        "by_user": {
            "terms": {
                "field": "user_id",
                "size": 2000,
                "min_doc_count": 1
            },
            "aggs": {
                "by_date": {
                    "date_histogram": {
                        "field": "timestamp",
                        "interval": "day",
                        "format": "yyyy-MM-dd-HH-mm"
                    },
                    "aggs": {
                        "unique_browsers": {
                            "terms": {
                                "field": "browser",
                                "size": 20,
                                "min_doc_count": 1
                            },
                            "aggs": {
                                "move_to_children": {
                                    "nested": {
                                        "type": "event"
                                    },
                                    "aggs": {
                                        "count_up": {
                                            "sum": "event.my-counter"
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
```

Now, if I remove the child piece and run it (with no cache), and then rerun - there is an instant speedup. If I add the child piece back, and run it multiple times, it always takes the same amount of time. Is it possible that is missing some form of cache layer? And if so, what are the reasons why/recommendations?

I ask because this aggregation is extremely slow (is this improved in 2.0?), it's taking ~60s with it as above, and if I remove it, the aggregations above take 500ms. I was hoping it'd be a first-hit-slow-second-hit-fast type deal like the other aggregations in use.

Basically, what are the best ways to make children aggregations faster? I've searched docs for about a week with no results on what I've tried, so I was wondering if I'm missing something, or if I should hold on for 2.0, if that does indeed help things.

Thanks in advance, let me know if more info is needed - not sure what is relevant. Hope I was clear :).

It should be noted we're talking about a lot of parents here (in the millions).

**Comment by jpountz** [comment]
parent/child is a bit tricky to cache: most of the things we cache are cached per segment but this doesn't work with parent child since the addition of a new document in segment A can change matching documents on segment B. So parent/child operations are never cached in elasticsearch 1.x. Elasticsearch 2.x makes it a bit better since there will be some caching happening if your index does not change too frequently but I wouldn't expect it to bring dramatic speedups.

**Comment by martijnvg** [comment]
@iwhitfield If you aggregation and data is relative static, you may want to check out the request cache, which can cache the entire shard level search response.

The reduce the time the first search request with the `children` agg after each refresh is taking, you can configure eager global ordinals loading on the `_parent` field.

In 2.0 for the `children` agg mainly the on heap memory footprint has significantly improved, because the paren/child data structures is doc values based.

**Comment by iwhitfield** [comment]
@jpountz @martijnvg thanks for the infos - what's the request cache, is that the query cache? If so, I have rolling time ranges in my filters so that probably wouldn't work, right?

To what @jpountz said, is it possible (because my indices are time based), that I can just merge old segments to have a single segment? Would that help, either currently or 2.0?

**Comment by adermlx** [comment]
i am fresher

**Comment by jpountz** [comment]
On 1.x there is nothing you can do, we explicitly prevent parent/child queries from being cached. On 2.x this should be possible if the index is (mostly) static, and merging everything down to a single segment will also help.

**Comment by iwhitfield** [comment]
@jpountz that's awesome, looking forward to 2.0. Even given the optimum scenario for use of the cache, should still not be expecting the same level of caching other aggregations receive?

**Comment by jpountz** [comment]
@iwhitfield sorry my last reply was more targeted at parent/child queries. For aggregations, you should already be able to get caching working if you use the query cache (renamed to request cache in 2.0 to better describe what it does). But this cache only works with search_type=COUNT and is invalidated everyt time there is a refresh operation happening.

**Comment by iwhitfield** [comment]
@jpountz ah, I can't do that because I've got moving time based filters in my queries sadly. 

So... I'm a little confused, because the caching & speed of has_child is fine in my testing, it's purely the children aggregation that's being a pain - is that just the nature of the join, and as good as it's going to get (i.e. never speeds up)?

**Comment by jpountz** [comment]
If it's very slow, any change that you can capture the node hot threads while the aggregation is running?

**Comment by iwhitfield** [comment]
@jpountz yeah I can try get that info for you, is there anything else which would be helpful?

I mentioned above that that set of aggregations (without the children aggregation) takes a little while to start with, but if I run it again it's milliseconds fast. I'm just trying to achieve the same, but the children aggregation is consistently the same amount of time regardless of how many times it has been run.

**Comment by jpountz** [comment]
> is there anything else which would be helpful?

Not really. This aggregation is known as slow and I suspect there is nothing that can be done about it. Just wanted to see the hot threads to confirm it's not something else.

**Comment by iwhitfield** [comment]
@jpountz here's a grab of the hot_threads API: https://gist.github.com/iwhitfield/ceefcccf83a4f20aa0de

**Comment by martijnvg** [comment]
@iwhitfield looking at your hot threads dump and it seems that a lot of time is spent at loading  parent/child field data and global ordinals. Have you tried configuring eager global ordinals loading?
(https://www.elastic.co/guide/en/elasticsearch/guide/current/parent-child-performance.html#_global_ordinals_and_latency)

**Comment by iwhitfield** [comment]
@martijnvg yeah, we have - here's my mapping for my child documents:

```
{
    "_all": {
        "enabled": false
    },
    "_parent": {
        "fielddata": {
            "loading": "eager_global_ordinals"
        },
        "type": "session"
    },
    "_routing": {
        "required": true
    },
    "_size": {
        "enabled": true
    },
    "_timestamp": {
        "enabled": true,
        "path": "timestamp"
    },
    "dynamic": "false",
    "properties": {
        "event": {
            "doc_values": true,
            "index": "not_analyzed",
            "type": "string"
        },
        "app": {
            "doc_values": true,
            "index": "not_analyzed",
            "type": "string"
        },
        "oid": {
            "doc_values": true,
            "index": "not_analyzed",
            "type": "string"
        },
        "ts": {
            "doc_values": true,
            "format": "dateOptionalTime",
            "type": "date"
        }
    }
}
```

Maybe I did something incorrectly?

**Comment by heiwen** [comment]
+1
